### PR TITLE
feature(hybrid-cloud): hybrid compatible auth middleware

### PR DESCRIPTION
**Pull request by corps**
Implements a hybrid cloud compatible AuthenticationMiddleware.

This implementation *has no effect on the behavior of existing SaaS and monolith deployments*!  It is 100% safe with no side effects.

When running in Silo modes, however, `request.user` becomes an `APIUser` instance and `request.auth` becomes an `AuthenticatedToken` instance.  Tests validate compatibility of existing and new authentication middleware.

### CODE REVIEW

File: src/sentry/services/hybrid_cloud/auth/impl.py [hunk 1]
```diff
@@ -114,3 +168,76 @@ def get_org_ids_with_scim(
...
+        return self.d.get(item, None) is not None
+
+    def get(self, key: str, default: str | bytes | None = None) -> str | bytes | None:
+        try:
+            return self[key]
+        except KeyError:
+            return default
+
+
+class FakeAuthenticationRequest:
```
Comment (mikejihbe) [inline src/sentry/services/hybrid_cloud/auth/impl.py:195]: Why is this necessary?

File: src/sentry/middleware/auth.py [hunk 2]
```diff
@@ -80,3 +102,43 @@ def process_exception(self, request: Request, exception):
...
 
             return expired(request, exception.user)
+
+
+class HybridCloudAuthenticationMiddleware(MiddlewareMixin):
+    def process_request(self, request: Request):
+        from sentry.web.frontend.accounts import expired
+
+        auth_result = auth_service.authenticate(request=authentication_request_from(request))
+        request.user_from_signed_request = auth_result.user_from_signed_request
```
Comment (markstory) [inline src/sentry/middleware/auth.py:112]: Longer term it would be nice to go back to one AuthenticationMiddleware and have this service handle the different logic required for monolith/silo mode.

File: src/sentry/middleware/auth.py [hunk 3]
```diff
@@ -80,3 +102,43 @@ def process_exception(self, request: Request, exception):
...
+        if auth_result.expired:
+            return expired(request, auth_result.user)
+        elif auth_result.user is not None:
+            request.user = auth_result.user
+            UserIP.log(auth_result.user, request.META["REMOTE_ADDR"])
+        else:
+            request.user = AnonymousUser()
+
+
+def authentication_request_from(request: Request) -> AuthenticationRequest:
```
Comment (markstory) [inline src/sentry/middleware/auth.py:125]: Should this be a method on the auth_service, or in the auth_service module?

File: src/sentry/services/hybrid_cloud/auth/impl.py [hunk 4]
```diff
@@ -76,8 +105,33 @@ def query_sso_state(
...
 
 class DatabaseBackedAuthService(AuthService):
-    # Monolith implementation that uses OrganizationMember and User tables freely, but in silo world
-    # this won't be possible.
+    def authenticate(self, *, request: AuthenticationRequest) -> AuthenticationResponse:
+        fake_request = FakeAuthenticationRequest(request)
+        handler: Any = RequestAuthenticationMiddleware()
+        expired_user: User | None = None
+        try:
+            handler.process_request(fake_request)
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/auth/impl.py:113]: What led to this approach? Shouldn't middleware being using the services? This looks like the service depending on middleware instead.

File: src/sentry/services/hybrid_cloud/auth/__init__.py [hunk 5]
```diff
@@ -66,3 +61,137 @@ class ApiAuthState:
...
+
+@dataclass
+class AuthenticationContext:
+    auth: AuthenticatedToken | None = None
+    user: APIUser | None = None
+
+    @contextlib.contextmanager
+    def applied_to_request(self, request: Any = None) -> Generator[None, None, None]:
+        """
+        Some code still reaches for the global 'env' object when determining user or auth behaviors.  This bleeds the
```
[the remark this answers is not in the document]
Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/__init__.py:146]: One, yes.  But, actually, I think this is still useful in the end.

Our app speaks request objects.  When we eventually implement RPC, we're going to need the ability to explicitly package up an authentication context and unpackage it into a request again.  *We will eventually need this code anyways*.  Doing it here, now, helps us start forcing our Stubs to simulate the idea that the request you come in with, isn't the request you necessary carry out via RPC.

For instance, using this logic, I was able to force stubs to create a fake context in which they are not authenticated, ensuring that stubs are not running logic that require user or request.auth context.  If they do, they must explicitly pass `auth_context` to make that clear.

If we don't take this approach, we must hunt down all the calls to `is_active_superuser` and replace them with... something else?  What will be better than the global request context in the end?

File: src/sentry/services/hybrid_cloud/auth/impl.py (hunk 1 above)
Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/impl.py:195]: Also, the other advantage of this request implementation is that it controls the inputs and outputs.

For instance, it does not allow session updates (the access is immutable) or clears.  Any mutation or side effects to the session must be explicitly modeled and not implicit.

The mutatable attributes are the only two that are allowed to "result" from authentication: `auth` and `user`.

Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/impl.py:195]: In fact, as designed, any usage within the authentication implementation of a session or header variable that does not exist will fail the test.  *only explicit inputs and outputs*, no implicit sharing of the originating request object.

File: src/sentry/middleware/auth.py (hunk 2 above)
Comment (corps) [reply src/sentry/middleware/auth.py:112]: In the future, Monolith mode implementation becomes deprecated and we basically only use HybridCloudAuthenticationMiddleware.  But this affords us safety by ensuring that this change *cannot impact the existing saas behavior*.  I felt authentication was the one case where splitting the implementation until we can fully vet this was the correct move.

File: src/sentry/middleware/auth.py (hunk 3 above)
Comment (corps) [reply src/sentry/middleware/auth.py:125]: Yes, in a follow up PR when I apply a similar approach to the `authentication_classes` of our json `Endpoints` I will do this.  Right now it was sensible to keep it local here.

File: src/sentry/middleware/auth.py [hunk 6]
```diff
@@ -35,12 +42,27 @@ def get_user(request):
                 )
                 user = AnonymousUser()
             else:
-                UserIP.log(user, request.META["REMOTE_ADDR"])
+                if SiloMode.get_current_mode() == SiloMode.MONOLITH:
+                    UserIP.log(user, request.META["REMOTE_ADDR"])
```
[the remark this answers is not in the document]
Comment (corps) [reply src/sentry/middleware/auth.py:46]: That's a prety excellent question that I went kind of back and forth on, I can take a closer look.

File: src/sentry/services/hybrid_cloud/auth/impl.py (hunk 4 above)
Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/impl.py:113]: In the end, the fact that `RequestAuthenticationMiddleware` is middleware, is just a convenience of its usage.  It's basically, *just a function that accepts and mutates a request object with a resulting user and auth value*.

So if it helps, I can drop the middleware part of this and make it a pure function instead.

Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/impl.py:113]: Ultimately, one way to think of this is, is that the control silo *is the definitive authority on authentication of a user*.  Ensuring that 100% of that logic is run inside of the service layer means that there is no risk of code drift from authenticating via an region silo.  Only the latest control silo providers authentication, and does so by providing the entirity of its encapsulated logic.

Service domains over service objects.

File: src/sentry/middleware/auth.py (hunk 2 above)
Comment (markstory) [reply src/sentry/middleware/auth.py:112]: That's fair. Reducing risk in the present is a good goal to have.

File: src/sentry/middleware/auth.py (hunk 3 above)
Comment (markstory) [reply src/sentry/middleware/auth.py:125]: Sounds good to me.

File: src/sentry/services/hybrid_cloud/auth/impl.py (hunk 4 above)
Comment (markstory) [reply src/sentry/services/hybrid_cloud/auth/impl.py:113]: >  Forcing that django logic to use the service means vendoring the entire django authentication stack (lots of functions, not just one or two).

Would what we need to do not fit within Django's [authentication backends](https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#writing-an-authentication-backend)? 

We could evolve our implementation towards a Django authentication backend in the future.

> Performance. A single RPC call that can capture all of the logic of authentication in the control silo was preferable to many many tiny little RPC calls sprinkled around for various objects.

That totally makes sense. Cross silo authentication will need to be a single RPC call in the future we can't take the performance overhead of multiple round trips.

Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/impl.py:113]: > Would what we need to do not fit within Django's [authentication backends](https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#writing-an-authentication-backend)?
We could evolve our implementation towards a Django authentication backend in the future.

We actually already do implement backends -- several of them.  Some of them return users, but some return social users.  some of them are intended as authentication and some of them are hammered in as suppport logic. Then we call back into django authentication helpers that assume a certain interface and implementation, which might include trying to fetch users on its own.  And then we do some more logic, like calling for Apitokens, etc, etc.   And the entry middelware, btw, has its own custom logic that won't fit into a backend altogether because of additional state and flow it carries. 

I, don't think we can re-use the django framework to solve this problem by the end.  We either put our service context outside  and above the spiderweb that is our authentication framework, and be forced to use a fake request object to perform the work, or we vendor and reimplement our own authentication framework.  Like, I've definitely tried otherwise and it didn't end well.

Comment (corps) [reply src/sentry/services/hybrid_cloud/auth/impl.py:113]: Oh, I also forgot how our authentication system currently uses *lazy user objets* to fit into the circular dependency with `Superuser`.  :///

File: src/sentry/services/hybrid_cloud/auth/impl.py [hunk 7]
```diff
@@ -114,3 +168,76 @@ def get_org_ids_with_scim(
...
+class FakeRequestDict:
+    d: Mapping[str, str | bytes | None]
+
+    def __init__(self, **d: Any):
+        self.d = d
+
+    def __getitem__(self, item: str) -> str | bytes:
+        result = self.d[item]
+        if result is None:
+            raise KeyError(f"Key '{item}' does not exist")
```
Comment (RyanSkonnord) [inline src/sentry/services/hybrid_cloud/auth/impl.py:182]: Nit: `f"Key {item!r} does not exist"` adds the quotation marks for you and handles escapes, if necessary.

**Review by RyanSkonnord — APPROVED** [review]
Per online discussion with @corps, I'm on board with the "fake request" approach. With the most recent comments added, I think this is clear enough to move forward and not taking on an undue amount of technical debt.
